src/java.base/share/classes/jdk/internal/math/FloatingDecimal.java

```diff
@@ -1826,726 +1824,619 @@ private static BinaryToASCIIConverter getBinaryToASCIIConverter(float f) {
-                    nTrailZero = 0;
-                } else if (c == '0') {
-                    digits[nDigits++] = (byte) c;
-                    nTrailZero++;
-                } else if (c == '.') {
-                    if (decSeen) {
-                        // already saw one ., this is the 2nd.
-                        throw new NumberFormatException("multiple points");
-                    }
-                    decPt = i;
-                    if (signSeen) {
-                        decPt -= 1;
-                    }
-                    decSeen = true;
-                } else {
-                    break digitLoop;
-                }
-                i++;
+    /**
+     * The input must match the {@link Double#valueOf(String) rules described here},
+     * about leading and trailing whitespaces, and the grammar.
+     *
+     * @param in the non-null input
+     * @param ix one of the {@code BINARY_<S>_IX} constants, where {@code <S>}
+     *          is one of 16, 32, 64
+     * @return an appropriate binary converter
+     * @throws NullPointerException if the input is null
+     * @throws NumberFormatException if the input is malformed
+     */
+    static ASCIIToBinaryConverter readJavaFormatString(String in, int ix) {
+        /*
+         * The scanning proper does not allocate any object,
+         * nor does it perform any costly computation.
+         * This means that all scanning errors are detected without consuming
+         * any heap, before actually throwing.
+         *
+         * Once scanning is complete, the method determines the length
+         * of a prefix of the significand that is sufficient for correct
+         * rounding according to roundTiesToEven.
+         * The actual value of the prefix length might not be optimal,
+         * but is always a safe choice.
+         *
+         * For hexadecimal input, the prefix is processed by this method directly,
+         * without allocating objects before creating the returned instance.
+         *
+         * For decimal input, the prefix is copied to the returned instance,
+         * along with the other information needed for the conversion.
+         * For comparison, the prefix length is at most
+         *       23 for BINARY_16_IX (Float16, once integrated in java.base)
+         *      114 for BINARY_32_IX (float)
+         *      769 for BINARY_64_IX (double)
+         */
+        int len = in.length();  // fail fast on null
+
+        /* Skip leading whitespaces. */
+        int i = skipWhitespaces(in, 0);  // main running index
+        if (i == len) {
+            throw new NumberFormatException("empty String");
+        }
+
+        /* Scan opt significand sign. */
+        int ch;  // running char
+        int ssign = ' ';  // ' ' iff sign is implicit
+        if ((ch = in.charAt(i)) == '-' || ch == '+') {  // i < len
+            ssign = ch;
+            ++i;
+        }
+
+        /*
+         * In some places the idiom
```

Comment:
Might be easier to read the code if this logic were refactored into a method.

Reply:
Right.
I'll take this note into account in the next phase of the cleanup.